edx/edx-enterprise
# [ENT-341] Setup enterprise API gateway & service endpoint for Programs

**Description:** This PR adds an Enterprise API service endpoint for Programs and Enterprise Catalogs which query course discovery, as well as an API gateway for those endpoints.

**JIRA:** [ENT-341](https://openedx.atlassian.net/browse/ENT-341)

**Merge deadline:** Coming Monday or slightly later.

**Testing instructions:**

Note: Anytime I say "send a GET request", the fastest way to do it is to just target the business sandbox through the browser (i.e. point to `https://business.sandbox.edx.org/enterprise/api/v1/enterprise-catalogs`). Make sure you're logged in as staff in the admin portal first, though.

- Testing `/enterprise/api/v1/enterprise-catalogs/`.

1. Send a GET request to `/enterprise/api/v1/enterprise-catalogs/` Note the results you get back.
1. Send a GET request to the discovery sandbox's `/api/v1/search/all/?q=` Note the results you get back.
1. Note that the results from the past 2 steps are equal.

- Testing `/enterprise/api/v1/enterprise-catalogs/{uuid}`.

1. Create an `EnterpriseCustomerCatalog` instance in the Django admin portal. Give it any query you want to test against (depending on the data available in the discovery sandbox). Note the UUID of the instance.
1. Send a GET request to `/enterprise/api/v1/enterprise-catalogs/{uuid}`, replacing `{uuid}` with the UUID you noted above. Note the results you get back.
1. Send a GET request to the discovery sandbox's `/api/v1/search/all/` with the query parameters equal to the query parameters you wrote in the instance. I.e. `/api/v1/search/all/?q=HelloWorld`. Note the results you get back.
1. Note that the results from the past 2 steps are equal.

- Testing `/enterprise/api/v1/programs/{uuid}`.

1. Send a GET request to `/enterprise/api/v1/programs/{uuid}`, replacing `{uuid}` with the UUID of an existing Program (search our discovery sandbox for some Program UUID, or create a new one and note the UUID). Note the results you get back.
1. Send a GET request to the discovery sandbox's `/api/v1/search/all/?q={uuid}` with `{uuid}` replaced with the Program UUID from above. Note the results you get back.
1. Note that the results from the past 2 steps are equal.

- Testing gateway

Note: I was given access to AWS by Matt Drayer. If you need access for any reason (although for testing this, you shouldn't unless you want to examine the actual gateway setup), ask him.

1. Get a JWT token from the business sandbox. You can do this using this cURL command:

`curl -H "Content-Type:application/x-www-form-urlencoded" -X POST -d "grant_type=client_credentials&client_id=<client-id>&client_secret=<client-secret>&token_type=jwt" "https://business.sandbox.edx.org/oauth2/access_token"`

Note to replace `<client-id>` and `<client-secret>` with the appropriate values (check for existing ones through the admin portal).

2. Use this JWT token in the following commands (replace `<jwt-token>` with the JWT token you generated above):

`curl -H "Authorization:JWT <jwt-token>" -X GET https://20enjrmi05.execute-api.us-west-2.amazonaws.com/edx_api/enterprise/v1/enterprise-catalogs/`
`curl -H "Authorization:JWT <jwt-token>" -X GET https://20enjrmi05.execute-api.us-west-2.amazonaws.com/edx_api/enterprise/v1/enterprise-catalogs/6b9f5528-23c9-4eff-aeb7-2755bf6780c1/`
`curl -H "Authorization:JWT <jwt-token>" -X GET https://20enjrmi05.execute-api.us-west-2.amazonaws.com/edx_api/enterprise/v1/programs/52ad909b-c57d-4ff1-bab3-999813a2479b/`

The above use UUIDs I setup before hand. You can play around by adding `EnterpriseCustomerCatalog`s and `Program`s as you like and use their UUIDs.

**Merge checklist:**

  **DO NOT** just add dependencies to `requirements/*.txt` files.

**Post merge:**
